saleor/order/base_calculations.py

```diff
@@ -19,95 +20,218 @@ def base_order_shipping(order: "Order") -> Money:
+            order,
+            subtotal,
+            undiscounted_subtotal,
+            shipping_price,
+            undiscounted_shipping_price,
+        )
+        subtotal_discount = undiscounted_subtotal - subtotal
+        apply_subtotal_discount_to_order_lines(
+            lines, undiscounted_subtotal, subtotal_discount
+        )
 
-def base_order_line_total(order_line: "OrderLine") -> OrderTaxedPricesData:
-    quantity = order_line.quantity
-    price_with_discounts = (
-        TaxedMoney(order_line.base_unit_price, order_line.base_unit_price) * quantity
+    return subtotal, shipping_price
+
+
+def apply_subtotal_discount_to_order_lines(
+    lines: Iterable["OrderLine"],
+    undiscounted_subtotal: Money,
+    subtotal_discount: Money,
+):
+    """Calculate order lines prices after applying discounts to entire subtotal."""
+    # Handle order with single line - propagate the whole discount to the single line.
+    lines = list(lines)
+    lines_count = len(lines)
+    if lines_count == 1:
+        line = lines[0]
+        apply_subtotal_discount_to_order_line(line, subtotal_discount)
+
+    # Handle order with multiple lines - propagate the order discount proportionally
+    # to the lines.
+    elif lines_count > 1:
+        remaining_discount = subtotal_discount
+        for idx, line in enumerate(lines):
+            if idx < lines_count - 1:
+                share = (
+                    line.base_unit_price_amount
+                    * line.quantity
+                    / undiscounted_subtotal.amount
+                )
+                discount = min(share * remaining_discount, undiscounted_subtotal)
+                apply_subtotal_discount_to_order_line(line, discount)
+                remaining_discount -= discount
+            else:
+                apply_subtotal_discount_to_order_line(line, remaining_discount)
+
+
+def apply_subtotal_discount_to_order_line(line: "OrderLine", discount: Money):
+    """Calculate order line prices after applying order level discount."""
+    currency = discount.currency
+    # This price includes line level discounts, but not entire order ones.
+    discounted_base_line_total = base_order_line_total(line).price_with_discounts.net
+    total_price = max(discounted_base_line_total - discount, zero_money(currency))
+    update_order_line_prices(line, total_price)
+
+
+def update_order_line_prices(line: "OrderLine", total_price: Money):
+    currency = total_price.currency
+    line.total_price_net = quantize_price(total_price, currency)
+    line.total_price_gross = line.total_price_net
+    line.undiscounted_total_price_gross_amount = (
+        line.undiscounted_total_price_net_amount
     )
-    undiscounted_price = (
-        TaxedMoney(
-            order_line.undiscounted_base_unit_price,
-            order_line.undiscounted_base_unit_price,
-        )
-        * quantity
+
+    quantity = line.quantity
+    if quantity > 0:
+        unit_price = total_price / quantity
```

Comment:
BTW, can you tell me why do we store rounded unit prices in db? We often do `unit_price * qunatity` in the codebase. To make the multiplication more accurate, unit prices shouldn't be rounded. We could round it only in resolvers and payloads. Is there any reason why we store rounded unit prices in db?

Reply:
> BTW, can you tell me why do we store rounded unit prices in db? We often do `unit_price * qunatity` in the codebase. To make the multiplication more accurate, unit prices shouldn't be rounded. We could round it only in resolvers and payloads. Is there any reason why we store rounded unit prices in db?

AFAIK unit_price is used only for display purposes. In case of re-calculations the base_price is used. unit-price also contains the taxes applied by external app/providers. It means that we could not be able to calculate the same total as with base_price.

Reply:
`base_price` is an unit price itself. We don't store `total_base_price`, but only `base_unit_price`. And all "base calculations" starts with `base_unit_price * quantity`, what you identified as anti-pattern. Check  `base_order_line_total` and `base_order_subtotal` functions for reference.

Reply:
Base_price defines the price calculated when creating the order. Used in any future recalculations. Unit price is taxed money which can be modified by tax plugins/app.
I didn't identify base-price*quantity as anti-patern. I said, that doing `total_line_price/quantity` could not give you the same result.
> So in the case of calculating the prices we should avoid flows like this one:
unit_price -> unit_price * quantity (total) -> total/quantity(unit_price)